pkg/cluster/sync.go

```diff
@@ -331,8 +335,17 @@ func (c *Cluster) syncSecrets() error {
 				userMap = c.pgUsers
 			}
 			pwdUser := userMap[secretUsername]
-			pwdUser.Password = string(curSecret.Data["password"])
-			userMap[secretUsername] = pwdUser
+			// if this secret belongs to the infrastructure role and the password has changed - replace it in the secret
+			if pwdUser.Password != string(curSecret.Data["password"]) && pwdUser.Origin == spec.RoleOriginInfrastructure {
+				c.logger.Debugf("updating the secret %q from the infrastructure roles", secretSpec.Name)
+				if _, err := c.KubeClient.Secrets(secretSpec.Namespace).Update(secretSpec); err != nil {
+					return fmt.Errorf("could not update infrastructure role secret for role %q: %v", secretUsername, err)
+				}
+			} else {
+				// for non-infrastructure role - update the role with the password from the secret
```

Comment:
1. So the operator is supposed to forcibly overwrite the passwords using secrets if the passwords change for whatever reason ?
2. Are non-infra-role passwords supposed to change w/o changing the relevant secrets ? If not, does this code branch deserve a warning log statement ?

Reply:
It is basically a feature for the infrastructure roles.

Reply:
The only reason for changing the role's password could be changes in the infrastructure roles secret. The idea behind these PR is to enable anyone who has corresponding permissions to change the password for an infrastructure role throughout the cluster by simply changing it in the infrastructure roles secret, providing a single place for setting passwords (and other properties as well). Before that PR, your changes of the password in the infrastructure role had no effect unless you remove the corresponding secret generated for that role.

Reply:
Answering question (2), the only user-definable non-infrastructure roles that have corresponding secrets are manifest roles, but we deliberately not allow to set passwords in the manifest. The only authoritative place to store those roles passwords are the corresponding secrets generated for them, so, contrary to the infrastructure roles, we don't have multiple places where we can set secrets and there is no need to sync them.